endpoints/openrtb2/auction.go

```diff
@@ -491,23 +491,41 @@ func fillAndValidateNativeAssets(assets []nativeRequests.Asset, impIndex int) er
 		return fmt.Errorf("request.imp[%d].native.request.assets must be an array containing at least one object", impIndex)
 	}
 
+	assetIDs := make(map[int64]struct{}, len(assets))
+
+	// If none of the asset IDs are defined by the caller, then prebid server should assign its own unique IDs. But
+	// if the caller did assign its own asset IDs, then prebid server will respect those IDs
+	assignAssetIDs := true
+	for i := 0; i < len(assets); i++ {
+		assignAssetIDs = assignAssetIDs && (assets[0].ID == 0)
+	}
+
 	for i := 0; i < len(assets); i++ {
-		// Per the OpenRTB spec docs, this is a "unique asset ID, assigned by exchange. Typically a counter for the array"
-		// To avoid conflict with the Request, we'll return a 400 if the Request _did_ define this ID,
-		// and then populate it as the spec suggests.
 		if err := validateNativeAsset(assets[i], impIndex, i); err != nil {
 			return err
 		}
-		assets[i].ID = int64(i)
+
+		if assignAssetIDs {
+			assets[i].ID = int64(i)
+			continue
+		}
+
+		// Each asset should have a unique ID thats assigned by the caller
+		if assets[i].ID == 0 {
```

Comment:
In case of us assigning the asset IDs, for the first asset we will always end up assigning the asset ID `0` as the value of `i` is going to be 0. We need to change: 
```
assets[i].ID = int64(i)
```
to

```
assets[i].ID = int64(i + 1)
```

So that we start assigning asset IDs from 1 and not 0.

Reply:
I just realized reading one of @SyntaxNode 's comment above that if the asset IDs are not specified we want to assign them as array indices and in that case `0` makes sense. Based on that we can leave this as is but I guess we can get rid of the next `if` check on L514, no? As an asset ID at that point of execution will be `0` only in two cases:
1. All asset IDs were `0` and we assigned the asset ID `0` for the first asset in which case it is valid
2. All asset IDs weren't `0`, meaning asset IDs were specified in the request itself and one of the assets have ID `0`. This should also be valid.

Reply:
we still need L514 imo, for partially defined asset IDs. L514 also shouldnt affect L509 since L509 is followed by a cont op

Reply:
e.g. if asset[0] has an ID but asset[1] doesnt

Reply:
Actually Mansi has a point. One asset with asset ID of 0 is valid. Multiple zeros are bad, but would be caught in the uniqueness check you have.

Reply:
well the issue is really that you can't tell if zero is defined by the caller or if its undefined, asset.ID isn't a pointer so it could be zero cuz of the caller or cuz its left out of the object

Reply:
I guess if we allow zero, its hard to figure out the appropriate error message

Reply:
I see what y'all are saying tho, its counter-intuitive to assign zero asset ID and then on the flip side not allow zero asset ID from the caller

Reply:
I'll update this by removing this check, so if multiple assets don't have an ID defined _or_ if multiple assets have zero as ID then the hashset check will catch it. I'll update the error message to appropriately explain that it could be either of two issues

Reply:
Sounds good to me.

Reply:
@Aadeshp Yeah you're right. I missed the `continue` statement. My bad. Even with that, as Hans mentioned one asset ID of `0` is still valid and I see you've already updated the PR for that. Thank you!